ext/drpttracer/drpttracer.c

```diff
@@ -0,0 +1,579 @@
+ * PT META DATA RELATED FUNCTIONS
+ */
+
+/* This function is used to get the PT's shared meta data:
+ *  cpu_family, cpu_model, cpu_step
+ */
+static void
+pt_shared_metadata_init(pt_metadata_t *metadata)
+{
+    memset(metadata, 0, sizeof(pt_metadata_t));
+    int cpu_family = 0, cpu_model = 0, cpu_stepping = 0;
+    get_cpu_info(&cpu_family, &cpu_model, &cpu_stepping);
+    metadata->cpu_family = (uint16_t)cpu_family;
+    metadata->cpu_model = (uint8_t)cpu_model;
+    metadata->cpu_stepping = (uint8_t)cpu_stepping;
+}
+
+/***************************************************************************
+ * PT TRACER HANDLE RELATED FUNCTIONS
+ */
+
+/* The type of PT trace's handle. The handle is used to store information about the
+ * current tracing. It contains the following fields:
+ *  1. The file descriptor of the perf event.
+ *  2. The perf event file's mmap pages.
+ *  3. The size of the perf event file's mmap pages.
+ *  4. The header of the perf event file's mmap page.
+ *  5. The ring buffer of the perf event's auxiliary data.
+ *
+ * The handle's data fields is transparent to the caller.
+ */
+typedef struct _pttracer_handle_t {
+    /* The file descriptor of the perf event. */
+    int fd;
+    /* The perf event file's mmap pages.
+     * The mmap pages stores the header page and the perf event's ring buffer.
+     */
+    void *base;
+    /* The size of the perf event file's mmap pages.
+     *  base_size = sizeof(header page) + sizeof(event ring buffer)
+     */
+    int base_size;
+    /* The header of the perf event file's mmap pages.
+     * The header is located at the beginning of the mmap pages.
+     * And it stores the all ring buffers' offset, size, head and tail.
+     * Also, it contains the perf event's meta data:
+     *  time_shift, time_mult, and time_zero.
+     */
+    struct perf_event_mmap_page *header;
+    /* The ring buffer of the perf event's auxiliary data.
+     *  sizeof(aux ring buffer) = header->aux_size.
+     */
+    void *aux;
+} pttracer_handle_t;
+
+/* The event ring buffer and aux ring buffer sizes must be set to (2^n)*_SC_PAGESIZE. We
+ * use the following macros to define these two buffers' size shifts.
+ */
+#define BASE_SIZE_SHIFT 8
+#define AUX_SIZE_SHIFT 8
+
+/* This function is used to generate a pttracer_handle_t from the perf event file
+ * descriptor.
+ */
+static pttracer_handle_t *
+pttracer_handle_create(IN void *drcontext, IN int fd)
+{
+    /* Initialize the perf event file's mmap pages. */
+    uint64_t base_size = (uint64_t)((1UL << BASE_SIZE_SHIFT) + 1) * sysconf(_SC_PAGESIZE);
+    uint64_t base_mmap_size = base_size;
+    void *base =
+        dr_map_file(fd, &base_mmap_size, 0, NULL, DR_MEMPROT_READ | DR_MEMPROT_WRITE, 0);
+    if (base == NULL || base_mmap_size != base_size) {
+        ERRMSG("failed to mmap perf event for pt tracing\n");
+        dr_close_file(fd);
+        return NULL;
+    }
+
+    /* Initialize the header of the perf event file's mmap pages. */
```

Comment:
Do we need to memset the first page to zero first?

Reply:
The page is read only to us. It is matained by perf in kernel. I modified the comment.